### PR TITLE
[ESP32] drivers: Add Watchdog Device tree support

### PR DESCRIPTION
**Pull request by ExtremeGTX**
- Add WDT(0,1) to esp32.dtsi
- Extend the module to be able to use WDT(0,1)
- Some minor refactoring due to usage of device tree

Tests:
- samples/drivers/watchdog
- tests/drivers/watchdog/wdt_basic_api

Note:
- timer module interrupt registers shall be removed when
timer driver implemented.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

### CODE REVIEW

File: soc/xtensa/esp32/dts_fixup.h [hunk 1]
```diff
@@ -8,4 +8,5 @@
 
 #define DT_CPU_CLOCK_FREQUENCY DT_CADENCE_TENSILICA_XTENSA_LX6_0_CLOCK_FREQUENCY
 #define CONFIG_ENTROPY_NAME DT_INST_0_ESPRESSIF_ESP32_TRNG_LABEL
+#define CONFIG_WDT_0_NAME DT_INST_0_ESPRESSIF_ESP32_WATCHDOG_LABEL
```
Comment (dcpleung) [inline soc/xtensa/esp32/dts_fixup.h]: `CONFIG_WDT_0_NAME` is only defined if `HAS_DTS_WDT` is not enabled. So there is no need to define this here.

File: drivers/watchdog/Kconfig.esp32 [hunk 2]
```diff
@@ -8,14 +8,23 @@
 menuconfig WDT_ESP32
 	bool "ESP32 Watchdog (WDT) Driver"
 	depends on SOC_ESP32
+	select HAS_DTS_WDT
 	default y
 	help
 	  Enable WDT driver for ESP32.
 
-config WDT_ESP32_IRQ
+config WDT0_ESP32_IRQ
```
Comment (dcpleung) [inline drivers/watchdog/Kconfig.esp32]: With DTS, you can use the config coming from the DTS, and there is no need for these IRQ kconfig anymore.

**Review by dcpleung — COMMENTED** [review]
Generally looks good. Just comments/questions on kconfig and DTS.

File: soc/xtensa/esp32/dts_fixup.h (hunk 1 above)
Comment (ExtremeGTX) [reply soc/xtensa/esp32/dts_fixup.h]: i think it should be replaced by `DT_WDT_0_NAME`, but without both tests don't compile

File: drivers/watchdog/Kconfig.esp32 (hunk 2 above)
Comment (dcpleung) [reply drivers/watchdog/Kconfig.esp32]: Ah... I thought DTS #define would still be there, but without the parent interrupt controller, the interrupt property cannot be set.

File: soc/xtensa/esp32/dts_fixup.h (hunk 1 above)
Comment (dcpleung) [reply soc/xtensa/esp32/dts_fixup.h]: Other boards simply do something like `#define DT_WDT_0_NAME DT_INST_0_ESPRESSIF_ESP32_WATCHDOG_LABEL` in their `dts_fixup.h`

**Review by dcpleung — APPROVED** [review]

File: drivers/watchdog/wdt_esp32.c [hunk 3]
```diff
@@ -22,135 +42,120 @@ struct wdt_esp32_data {
...
+	((volatile struct wdt_esp32_regs_t  *)(DEV_CFG(dev))->base)
 
 /* ESP32 ignores writes to any register if WDTWPROTECT doesn't contain the
  * magic value of TIMG_WDT_WKEY_VALUE.  The datasheet recommends unsealing,
  * making modifications, and sealing for every watchdog modification.
  */
-static inline void wdt_esp32_seal(void)
+static inline void wdt_esp32_seal(struct device *dev)
 {
-	volatile u32_t *reg = (u32_t *)TIMG_WDTWPROTECT_REG(1);
```
Comment (lpereira) [inline drivers/watchdog/wdt_esp32.c]: Any reason to not use the macros from the ESP32 SDK?

File: dts/bindings/watchdog/espressif,esp32-watchdog.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2018, Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+title: Nordic Semiconductor NRF watchdog driver
```
Comment (lpereira) [inline dts/bindings/watchdog/espressif,esp32-watchdog.yaml]: Shouldn't this say ESP32?

File: drivers/watchdog/wdt_esp32.c (hunk 3 above)
Comment (ExtremeGTX) [reply drivers/watchdog/wdt_esp32.c]: - It is more easier to manage device dependent registers using `dev` instance. 
- I found it more readable and modular to group everything related to registers addresses and offset in config struct
- I used the macros in config struct [lines 242,243](https://github.com/zephyrproject-rtos/zephyr/pull/20265/files#diff-6dc0427c13d71084aba1bc052bfbe136R242-R243)
- I'm trying to be independent of esp-idf as much as i can, so we gain more control over our code.

Correct me if i'm wrong or missed any thing

Comment (galak) [reply drivers/watchdog/wdt_esp32.c]: @lpereira any feedback?

Comment (lpereira) [reply drivers/watchdog/wdt_esp32.c]: It's kind of hard to get rid of ESP-IDF because of the HAL library, but this is fine too.

**Pull request merged**